[Beta]: docs(cn): responding-to-events.md (#635)
reactjs/zh-hans.reactjs.org: beta/src/pages/learn/responding-to-events.md

---
title: 响应事件
translators:
  - Jiacheng787
---

<Intro>

> React 允许你在 JSX 中添加事件处理函数。事件处理函数是你自定义的函数，它将在响应用户交互（如点击、悬停、表单输入框获得焦点等）时触发。

Reviewer: ```suggestion
使用 React 可以在 JSX 中添加事件处理函数。其中事件处理函数为自定义函数，它将在响应用户交互（如点击、悬停、表单输入框获得焦点等）时触发。
```

Reviewer: 这个为什么关了？
加个其中会让语句更连贯，你有什么疑义，得说出来，别直接关闭...

Author: 好的，我就是觉得这种主观性比较强，还是按照原文的表达更好

Reviewer: 那你得说明一下啊，不然我不知道发生了什么就关了...

Author: 好的，顺便问下啥时候merge

Reviewer: 还没看完，看完 merge

</Intro>

<YouWillLearn>

* 编写事件处理函数的不同方法
* 如何从父组件传递事件处理逻辑
* 事件如何传播以及如何停止它们

</YouWillLearn>

## 添加事件处理函数 {#adding-event-handlers}

> 要添加一个事件处理函数，你需要先定义一个函数，然后 [将它作为 prop 传入](/learn/passing-props-to-a-component) 合适的 JSX 标签。例如，这是一个尚未执行任何操作的按钮：

Reviewer: ```suggestion
如需添加一个事件处理函数，你需要先定义一个函数，然后 [将其作为 prop 传入](/learn/passing-props-to-a-component) 合适的 JSX 标签。例如，这里有一个没绑定任何事件的按钮：
```

<Sandpack>

```js
export default function Button() {
  return (
    <button>

>       我什么都不做

Reviewer: ```suggestion
      未绑定任何事件
```

    </button>
  );
}
```

</Sandpack>

你可以通过以下三个步骤使它在用户点击时显示消息：

1. 在 `Button` 组件 *内部* 声明一个名为 `handleClick` 的函数。
2. 实现函数内部的逻辑（使用 `alert` 来显示消息）。

> 3. 添加 `onClick={handleClick}` 到 `<button>` JSX 。

Reviewer: ```suggestion
3. 添加 `onClick={handleClick}` 到 `<button>` JSX 中。
```

|----------------------------------------|--------------------------------|
| `<button onClick={handleClick}>` | `<button onClick={handleClick()}>` |

区别很微妙。在第一个示例中，`handleClick` 函数作为 `onClick` 事件处理函数传递。这会让 React 记住它，并且只在用户点击按钮时调用你的函数。

在第二个示例中，`handleClick()` 中最后的 `()` 会在 [渲染](/learn/render-and-commit) 过程中 *立即* 触发函数，即使没有任何点击。这是因为在 [JSX `{` 和 `}`](/learn/javascript-in-jsx-with-curly-braces) 之间的 JavaScript 会立即执行。

当你编写内联代码时，同样的陷阱可能会以不同的方式出现：

| 传递一个函数（正确）           | 调用一个函数（错误） |
|----------------------------------------|--------------------------------|
| `<button onClick={() => alert('...')}>` | `<button onClick={alert('...')}>` |

如果按如下方式传递内联代码，并不会在点击时触发，而是会在每次组件渲染时触发：

```jsx
// 这个 alert 在组件渲染时触发，而不是点击时触发！
<button onClick={alert('你点击了我！')}>
```

如果你想要定义内联事件处理函数，请将其包装在匿名函数中，如下所示：

```jsx
<button onClick={() => alert('你点击了我！')}>
```

这里创建了一个稍后调用的函数，而不会在每次渲染时执行其内部代码。

在这两种情况下，你要传递的是一个函数：

* `<button onClick={handleClick}>` 传递了 `handleClick` 函数。
* `<button onClick={() => alert('...')}>` 传递了 `() => alert('...')` 函数。

> 查看 [JavaScript Refresher](a-javascript-refresher#arrow-functions) 了解更多箭头函数的信息。

</Gotcha>

### 在事件处理函数中读取 props {/*reading-props-in-event-handlers*/}

> 因为事件处理函数是在组件内部声明的，所以他们可以访问组件的 props 。下面是一个按钮，在点击时会显示带有其 `message` prop 的 alert ：

Reviewer: ```suggestion
由于事件处理函数是声明于组件内部，因此他们可以直接访问组件的 props。示例中的按钮，当点击时会弹出带有 `message` prop 的 alert：
```

</Sandpack>

> 你定义了 `handleClick` 函数然后 [将它作为 prop 传入](/learn/passing-props-to-a-component) `<button>` 。`handleClick` 是一个 **事件处理函数** 。事件处理函数:

Reviewer: ```suggestion
你可以定义 `handleClick` 函数然后 [将其作为 prop 传入](/learn/passing-props-to-a-component) `<button>`。其中 `handleClick` 是一个 **事件处理函数** 。事件处理函数通常有如下特点:
```

</Sandpack>

此处有两个按钮，会展示不同的消息。你可以尝试更改传递给他们的消息。

### 将事件处理函数作为 props 传递 {/*passing-event-handlers-as-props*/}

> 通常，你会希望让父组件指定子组件的事件处理函数。考虑按钮：根据你使用 `Button` 组件的位置，你可能想要执行不同的功能——也许一个播放电影，另一个上传图片。

Reviewer: ```suggestion
通常，我们会在父组件中定义子组件的事件处理函数。比如：置于不同位置的 `Button` 组件，可能最终执行的功能也不同 —— 也许是播放电影，也许是上传图片。
```

> 为此，将组件从其父组件接收的 prop 作为事件处理函数传递，如下所示：

Reviewer: ```suggestion
为此，将组件从父组件接收的 prop 作为事件处理函数传递，如下所示：
```

</Sandpack>

> 在这里，`Toolbar` 组件渲染了一个 `PlayButton` 组件和 `UploadButton` 组件：

Reviewer: ```suggestion
示例中，`Toolbar` 组件渲染了一个 `PlayButton` 组件和 `UploadButton` 组件：
```

- `PlayButton` 将 `handlePlayClick` 作为 `onClick` prop 传入 `Button` 组件内部。
- `UploadButton` 将 `() => alert('正在上传！')` 作为 `onClick` prop 传入 `Button` 组件内部。

最后，你的 `Button` 组件接收一个名为 `onClick` 的 prop 。它直接将这个 prop 以 `onClick={onClick}` 方式传递给浏览器内置的 `<button>` 。这告诉 React 在点击时调用传入的函数。

> 如果你使用一个 [设计系统](https://uxdesign.cc/everything-you-need-to-know-about-design-systems-54b109851969)，按钮等组件通常包含样式但不指定行为。相反，类似 `PlayButton` 和 `UploadButton` 的组件会向下传递事件处理函数。

Reviewer: ```suggestion
如果你遵循某个 [设计系统](https://uxdesign.cc/everything-you-need-to-know-about-design-systems-54b109851969) 时，类似于按钮这类的组件，通常会包含样式但不指定行为。而类似 `PlayButton` 和 `UploadButton` 这类的组件则会向下传递事件处理函数。
```

### 命名事件处理函数 prop {/*naming-event-handler-props*/}

内置组件例如 `<button>` 和 `<div>` 仅支持 [浏览器事件名称](/reference/reactdom-api) ，例如 `onClick` 。但是，当你构建自己的组件时，你可以按你喜欢的任何方式命名它们的事件处理函数 props 。

> 按照惯例，事件处理函数 props 应该以 `on` 开头，后跟一个大写字母。

> 例如，这个 `Button` 组件的 `onClick` prop 可以被命名为 `onSmash` ：

Reviewer: ```suggestion
例如，`Button` 组件的 `onClick` prop 可以被命名为 `onSmash`：
```

</Sandpack>

在这里例子中，`<button onClick={onSmash}>` 表明浏览器内置的 `<button>`（小写）仍然需要一个 `onClick` prop ，但是你自定义的 `Button` 组件接收到的 prop 名称由你决定！

> 当你的组件支持多种交互时，你可以用应用程序特定的概念命名事件处理函数 prop 。例如，一个 `Toolbar` 组件接收 `onPlayMovie` 和 `onUploadImage` 两个事件处理函数：

Reviewer: ```suggestion
当你的组件支持多种交互时，你可以根据不同的应用程序命名事件处理函数 prop。例如，一个 `Toolbar` 组件接收 `onPlayMovie` 和 `onUploadImage` 两个事件处理函数：
```

</Sandpack>

> 请注意 `App` 组件不需要知道 `Toolbar` 将会对 `onPlayMovie` 和 `onUploadImage` 做 *什么* 。上面是一个 `Toolbar` 的实现细节。在这里，`Toolbar` 将它们作为 `onClick` 处理函数向下传递给它的 `Button` ，但也可以稍后通过快捷键触发它们。用应用程序特定的交互方式（如 `onPlayMovie`）来命名 prop ，可以让你灵活地更改以后使用它们的方式。

Reviewer: ```suggestion
请注意，`App` 组件并不需要知道 `Toolbar` 将会对 `onPlayMovie` 和 `onUploadImage` 做 **什么** 。上述示例是 `Toolbar` 的实现细节。其中，`Toolbar` 将它们作为 `onClick` 处理函数传递给了 `Button` 组件，其实还可以通过键盘快捷键来触发它们。根据应用程序特定的交互方式（如 `onPlayMovie`）来命名 prop ，可以让你灵活地更改以后使用它们的方式。
```

## 事件传播 {/*event-propagation*/}

<!--
// TODO illo
-->

> 事件处理函数还将捕获来自组件可能拥有的任何子组件的事件。我们说一个事件“冒泡”或“传播”到树的上方：它从事件发生的地方开始，然后沿着树向上传播。

Reviewer: ```suggestion
事件处理函数还将捕获任何来自子组件的事件。通常，我们会说事件“冒泡”或“传播”到树的上方：它从事件发生的地方开始，然后沿着树向上传播。
```

> 下面这个 `<div>` 包含两个按钮。`<div>` *和* 每个按钮都有自己的 `onClick` 处理函数。你认为点击按钮时会触发哪些处理函数？

Reviewer: ```suggestion
下面这个 `<div>` 包含两个按钮。`<div>` **和** 每个按钮都有自己的 `onClick` 处理函数。你认为点击按钮时会触发哪些处理函数？
```

</Sandpack>

> 如果你点击任一按钮，它自身的 `onClick` 将首先执行，然后是父级 `<div>` 的 `onClick` 。所以会出现两条消息。如果你点击 toolbar 本身，只有父级 `<div>` 的 `onClick` 会执行。

Reviewer: ```suggestion
如果你点击任一按钮，它自身的 `onClick` 将首先执行，然后是父级 `<div>` 的 `onClick`。因此会出现两条消息。如果你点击 toolbar 本身，将只有父级 `<div>` 的 `onClick` 会执行。
```

<Gotcha>

在 React 中所有事件都会传播，除了 `onScroll` ，它仅适用于你附加到的 JSX 标签。

</Gotcha>

> ### 停止传播 {/*event-propagation*/}

Reviewer: ```suggestion
### 停止传播 {/*stopping-propagation*/}
```

> ### 停止传播 {/*stopping-propagation*/}

Reviewer: ```suggestion
### 阻止传播 {/*stopping-propagation*/}
```

> 事件处理函数接收一个 **事件对象** 作为它们唯一的参数。按照惯例，它通常被称为 `e` ，代表 "event"（事件）。你可以使用此对象来读取有关事件的信息。

Reviewer: ```suggestion
事件处理函数接收一个 **事件对象** 作为唯一的参数。按照惯例，它通常被称为 `e` ，代表 "event"（事件）。你可以使用此对象来读取有关事件的信息。
```

> 这个事件对象还允许你停止传播。如果你想阻止一个事件到达父组件，你需要像下面 `Button` 组件那样调用 `e.stopPropagation()` ：

Reviewer: ```suggestion
这个事件对象还允许你阻止传播。如果你想阻止一个事件到达父组件，你需要像下面 `Button` 组件那样调用 `e.stopPropagation()` ：
```

</Sandpack>

当你点击按钮时：

1. React 调用了传递给 `<button>` 的 `onClick` 处理函数。
2. 定义在 `Button` 中的处理函数执行了如下操作：

>    * 调用 `e.stopPropagation()` ，防止事件进一步冒泡。

Reviewer: ```suggestion
   * 调用 `e.stopPropagation()`，防止事件进一步冒泡。
```

>    * 调用 `onClick` 函数，它是从 `Toolbar` 组件传递过来的 prop 。

Reviewer: ```suggestion
   * 调用 `onClick` 函数，它是从 `Toolbar` 组件传递过来的 prop。
```

> 3. 在 `Toolbar` 组件中定义的函数，显示按钮对应的 alert 。

Reviewer: ```suggestion
3. 在 `Toolbar` 组件中定义的函数，显示按钮对应的 alert。
```

> 4. 由于传播被停止，父级 `<div>` 的 `onClick` 处理函数不会执行。

Reviewer: ```suggestion
4. 由于传播被阻止，父级 `<div>` 的 `onClick` 处理函数不会执行。
```

> 由于 `e.stopPropagation()` ，点击按钮现在仅显示一个 alert（来自 `<button>`）而不是其中的两个（来自 `<button>` 和父级 toolbar `<div>`）。点击按钮与点击周围的 toolbar 不同，因此停止传播对这个 UI 是有意义的。

Reviewer: ```suggestion
由于调用了 `e.stopPropagation()`，点击按钮现在将只显示一个 alert（来自 `<button>`），而并非两个（分别来自 `<button>` 和父级 toolbar `<div>`）。点击按钮与点击周围的 toolbar 不同，因此阻止传播对这个 UI 是有意义的。
```

<DeepDive title="捕获阶段的事件">

<!--
// TODO Illo
-->

> 极少数情况下，你可能需要捕获子元素上的所有事件，*即使它们停止了传播* 。例如，你可能希望将每次点击记录到分析中，而不管传播逻辑是什么。你可以通过在事件名称末尾添加 `Capture` 来实现：

Reviewer: ```suggestion
极少数情况下，你可能需要捕获子元素上的所有事件，**即便它们阻止了传播**。例如，你可能希望将每次点击进行埋点记录中，而不管传播逻辑是什么。你可以通过在事件名称末尾添加 `Capture` 来实现：
```

```js
<div onClickCapture={() => { /* 这会首先执行 */ }}>
  <button onClick={e => e.stopPropagation()} />
  <button onClick={e => e.stopPropagation()} />
</div>
```

每个事件分三个阶段传播：

1. 它向下传播，调用所有的 `onClickCapture` 处理函数。
2. 它执行被点击元素的 `onClick` 处理函数。 
3. 它向上传播，调用所有的 `onClick` 处理函数。

捕获事件对于路由器或分析等代码很有用，但你可能不会在应用程序代码中使用它们。

</DeepDive>

### 传递处理函数作为传播的替代方案 {/*passing-handlers-as-alternative-to-propagation*/}

> 注意这个点击事件处理函数如何执行一行代码 _然后_ 调用父组件传递的 `onClick` prop ：

Reviewer: ```suggestion
注意，此处的点击事件处理函数先执行了一行代码，**然后**调用了父组件传递的 `onClick` prop：
```

}
```

> 你也可以在调用父元素 `onClick` 函数之前向这个处理函数添加更多代码。此模式为传播提供了一个 *替代方案* 。它让子组件处理事件，同时也让父组件指定一些额外的行为。与传播不同，它不是自动的。但是这种模式的好处是，你可以清楚地跟踪作为某个事件执行的整个调用链。

Reviewer: ```suggestion
你也可以在调用父元素 `onClick` 函数之前，向这个处理函数添加更多代码。此模式是事件传播的另一种 *替代方案* 。它让子组件处理事件，同时也让父组件指定一些额外的行为。与事件传播不同，它并非自动。但是使用这种模式的好处是，你可以清楚地追踪某个事件的结果，以及整个调用链的执行。
```

> 你也可以在调用父元素 `onClick` 函数之前，向这个处理函数添加更多代码。此模式是事件传播的另一种 *替代方案* 。它让子组件处理事件，同时也让父组件指定一些额外的行为。与事件传播不同，它并非自动。但是使用这种模式的好处是，你可以清楚地追踪某个事件的结果，以及整个调用链的执行。

Reviewer 2: “你可以清楚地追踪因某个事件的触发而执行的整条代码链”

句子的主干是 track the code chain，其余的都是定语

> 如果你依赖于传播，而且很难追踪哪些处理程序在执行，以及执行的原因，可以试试这种方法。

Reviewer: ```suggestion
如果你依赖于事件传播，而且很难追踪哪些处理程序在执行，及其执行的原因，可以尝试这种方法。
```

### 阻止默认行为 {/*preventing-default-behavior*/}

> 某些浏览器事件具有与其关联的默认行为。例如，点击 `<form>` 表单内部的按钮会触发表单提交事件，默认情况下将重新加载整个页面：

Reviewer: ```suggestion
某些浏览器事件具有与事件相关联的默认行为。例如，点击 `<form>` 表单内部的按钮会触发表单提交事件，默认情况下将重新加载整个页面：
```

</Sandpack>

> 你可以调用事件对象上的 `e.preventDefault()` 来阻止这种情况发生：

Reviewer: ```suggestion
你可以调用事件对象中的 `e.preventDefault()` 来阻止这种情况发生：
```

</Sandpack>

> 不要混淆 `e.stopPropagation()` 和 `e.preventDefault()` 。它们都很有用，但是不相关：

Reviewer: ```suggestion
不要混淆 `e.stopPropagation()` 和 `e.preventDefault()`。它们都很有用，但是不相关：
```

> * [`e.stopPropagation()`](https://developer.mozilla.org/docs/Web/API/Event/stopPropagation) 停止触发附加到标签之外的事件处理函数。

Reviewer 2: 阻止绑定在上方标签的事件处理函数触发

> * [`e.preventDefault()` ](https://developer.mozilla.org/docs/Web/API/Event/preventDefault) 阻止少数事件的默认浏览器行为。

Reviewer: ```suggestion
* [`e.preventDefault()`](https://developer.mozilla.org/docs/Web/API/Event/preventDefault) 阻止少数事件的默认浏览器行为。
```

## 事件处理函数可以包含副作用吗？ {/*can-event-handlers-have-side-effects*/}

> 当然可以！事件处理函数是产生副作用的最佳位置。

Reviewer 2: 建议把“产生”改为“保存”或者“使用”

> 当然了！事件处理函数是产生副作用的最佳位置。

Reviewer: ```suggestion
当然可以！事件处理函数是产生副作用的最佳位置。
```

> 与渲染函数不同，事件处理函数不需要是 [纯函数](/learn/keeping-components-pure)，因此它是 *更改* 某些内容的绝佳之选。例如，更改输入框的值以响应键入，或者更改列表以响应按钮的触发。但是，为了更改某些信息，你首先需要某种方式存储它。在 React 中，这是通过 [state（�组件的缓存）](/learn/state-a-components-memory) 来完成的。你将在下一章节了解所有相关信息。

Reviewer 2: “因此它是 *更改* 某些内容的绝佳之选” place 没有翻译出来，建议改为“因此它是 *更改* 内容的绝佳位置”

> 与渲染函数不同，事件处理函数不需要是 [纯函数](/learn/keeping-components-pure) ，因此它是 *更改* 某些内容的好地方——例如，更改输入值以响应键入，或者更改列表以响应按钮按下。但是，为了更改某些信息，你首先需要某种方式存储它。在 React 中，这是通过 [state, a component's memory](/learn/state-a-components-memory) 来完成的。你将在下一页了解所有相关信息。

Reviewer: ```suggestion
与渲染函数不同，事件处理函数不需要是 [纯函数](/learn/keeping-components-pure)，因此它是 *更改* 某些内容的绝佳之选。例如，更改输入框的值以响应键入，或者更改列表以响应按钮的触发。但是，为了更改某些信息，你首先需要某种方式存储它。在 React 中，这是通过 [state（组件的缓存）](/learn/state-a-components-memory) 来完成的。你将在下一章节了解所有相关信息。
```

<Recap>

* 你可以通过将函数作为 prop 传递给元素如 `<button>` 来处理事件。

> * 必须传递事件处理函数，**而不是调用！** `onClick={handleClick}` ，不是 `onClick={handleClick()}` 。

Reviewer: ```suggestion
* 必须传递事件处理函数，**而非函数调用！** `onClick={handleClick}` ，不是 `onClick={handleClick()}`。
```

* 你可以单独或者内联定义事件处理函数。

> * 事件处理函数在组件内部定义，所以它们可以访问 props 。

Reviewer: ```suggestion
* 事件处理函数在组件内部定义，所以它们可以访问 props。
```

> * 你可以在父组件中定义一个事件处理函数，并将其作为 prop 传递给子元素。

Reviewer: ```suggestion
* 你可以在父组件中定义一个事件处理函数，并将其作为 prop 传递给子组件。
```

> * 你可以使用应用程序特定的名称定义自己的事件处理函数 prop 。

Reviewer: ```suggestion
* 你可以根据特定于应用程序的名称定义事件处理函数的 prop。
```

> * 事件向上传播。通过事件的第一个参数调用 `e.stopPropagation()` 来防止这种情况。

Reviewer 2: 事件**会**向上传播。

> * 事件向上传播。在第一个参数上调用 `e.stopPropagation()` 来防止这种情况。

Reviewer: ```suggestion
* 事件向上传播。通过事件的第一个参数调用 `e.stopPropagation()` 来防止这种情况。
```

* 事件可能具有不需要的浏览器默认行为。调用 `e.preventDefault()` 来阻止这种情况。

> * 从子组件显式调用事件处理函数 prop 是传播的一个很好的替代方案。

Reviewer: ```suggestion
* 从子组件显式调用事件处理函数 prop 是事件传播的另一种优秀替代方案。
```

</Recap>

<Challenges>

> ### 修复一个事件处理函数 {/*fix-an-event-handler*/}

Reviewer: ```suggestion
### 修事件处理函数 {/*fix-an-event-handler*/}
```

Author: 改成 “修复事件处理函数” 是否更好？

Reviewer: 太卡了，误删了，我重启下电脑

> 点击这个按钮应该可以在白色和黑色之间切换页面背景。然而，当你点击它时，什么也没有发生。解决这个问题。（不用担心 `handleClick` 里面的逻辑——那部分很好。）

Reviewer: ```suggestion
点击此按钮理论上应该在黑白主题之间切换页面背景。然而，当你点击它时，什么也没有发生。解决这个问题。（无需担心 `handleClick` 的内部逻辑。）
```

<Solution>

> 这是由于 `<button onClick={handleClick()}>` 在渲染过程中 _调用_ 了 `handleClick` 函数而不是 _传递_ 它。移除 `()` 调用改为 `<button onClick={handleClick}>` 进而修复问题：

Reviewer: ```suggestion
这是由于 `<button onClick={handleClick()}>` 在渲染过程中 _调用_ 了 `handleClick` 函数，而没有将其进行 _传递_。移除 `()` 调用改为 `<button onClick={handleClick}>` 进而修复问题：
```

</Solution>

> ### 连接事件 {/*wire-up-the-events*/}

Reviewer: ```suggestion
### 关联事件 {/*wire-up-the-events*/}
```

> 这个 `ColorSwitch` 组件渲染一个按钮。它应该改变页面颜色。将它与从父组件接收的 `onChangeColor` 事件处理函数 prop 连接，以便在点击按钮时改变颜色。

Reviewer: ```suggestion
`ColorSwitch` 组件渲染了一个按钮。它应该改变页面颜色。将它与从父组件接收的 `onChangeColor` 事件处理函数关联，以便在点击按钮时改变颜色。
```

> 在你这样做了之后，你会发现点击按钮也会增加页面点击计数器的值。编写父组件的同事坚持 `onChangeColor` 不会增加计数器的值。还发生了什么呢？修复它，使点击按钮 *只* 改变颜色，并且 _不_ 增加计数器。

Reviewer: ```suggestion
如此操作后，你会发现点击按钮时，也会增加页面点击计数器的值。而编写父组件的同事认为 `onChangeColor` 不应该增加计数器的值。应该如何处理？如何修改才能使点击按钮 **只** 改变颜色，并且 **不** 增加计数器。
```

Author: 改为下面这样看看会不会更好：

“如此操作后，你会发现点击按钮时，也会增加页面点击计数器的值。而编写父组件的同事坚持认为 `onChangeColor` 不会增加计数器的值。还发生了什么事情呢？修复问题使点击按钮 **只** 改变颜色，并且 **不** 增加计数器。”

Reviewer: 调整了，这里 `还发生了什么事情呢？` 感觉很奇怪

<Solution>

> 首先，你需要添加事件处理函数，例如 `<button onClick={onChangeColor}>` 。

Reviewer: ```suggestion
首先，你需要添加事件处理函数，例如 `<button onClick={onChangeColor}>`。
```

> 但是，这里引入了增加计数器的问题。如果正如你的同事所坚持的那样，`onChangeColor` 不这样做，那么问题在于这个事件向上传播，并且上面的某些事件处理函数会这样做。为了解决这个问题，你需要停止传播。但是不要忘记你仍然应该调用 `onChangeColor` 。

Reviewer: ```suggestion
然而，同时又引入了计数器递增的问题。正如你的同事所坚持的那样，`onChangeColor` 并不符合预期，这主要是因为事件发生向上传播，并且上层的某些事件处理函数执行了递增操作。为了解决这个问题，你需要阻止事件冒泡。但是不要忘记调用 `onChangeColor`。
```
